Rewrite get_default_modules()

sgallagher opened the pull request:
The previous version was needlessly complicated and difficult to
follow. This patch replaces it with a lookup of the latest
versions of each of the default streams and a recursive lookup of
their dependencies.

Also adds a `--debug` flag and some debugging messages.

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

sgallagher committed:
Rewrite get_default_modules()

The previous version was needlessly complicated and difficult to
follow. This patch replaces it with a lookup of the latest
versions of each of the default streams and a recursive lookup of
their dependencies.

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

Conan-Kudo commented on splitter.py:
```diff
@@ -184,6 +185,70 @@ def validate_filenames(directory, repoinfo):
+            # Ignore "platform" because it's special
+            if modname == "platform":
+                logging.debug('Skipping platform')
+                continue
```
Is `platform` the only case we want to ignore?

sgallagher replied:
`platform` is the only special-case for runtime. If that ever changes, we'll need to modify this (but we'd also need to modify it in many other projects/libraries)

Conan-Kudo commented:
@sgallagher I might be reading this a bit naively, but it seems like this code assumes that the modules and module dependencies are resolvable within a repository. Is that a correct assumption?

sgallagher commented:
> @sgallagher I might be reading this a bit naively, but it seems like this code assumes that the modules and module dependencies are resolvable within a repository. Is that a correct assumption?

It is working under the assumption that we are interacting with the RHEL 8 AppStream repository, which contains all modules shipped by RHEL 8.

If we want to make a different assumption, then I guess I'd rework the patch to just ensure we get whatever subset of the streams are available in this repository.

Conan-Kudo commented:
@sgallagher I think for now this is fine, but I'm more concerned about dealing with layered repositories that contain modules that describe module dependencies in base layer repositories.

sgallagher commented:
@Conan-Kudo Yeah, if we want this to be reusable for general purposes (such as working with Fedora updates repos), we should probably make the change I mentioned above. But if we only care about RHEL 8, this is probably fine.

@smooge Can I get your opinion on this matter?

sgallagher commented:
I pushed two minor changes (amended). The first was to fix up a couple formatting issues and the second changes how `_get_recursive_dependencies()` saves its results back to the `all_deps` variable. Previously, if we ended up skipping a default stream because its dependencies didn't resolve, we were still including whatever subset was available. Now we exclude that, so if we skip a default stream we don't carry it along.

The previous behavior was safe, just potentially results in carrying more content than is desired.

sgallagher committed:
Add --ignore-missing-default-deps argument

There may be cases where GrobiSplitter needs to get the default
streams from a repository where that repository does not contain
all of those streams' dependencies. (Such as 'fedora-updates',
which may depend on streams that are only in the 'fedora' stable
repository.

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

sgallagher commented:
And one more change, this one in its own patch. It adds a `--ignore-missing-default-deps` argument that will allow us to retrieve as much as possible for the default streams, even if some of their dependencies are coming from another repository.

@Conan-Kudo Would you mind re-reviewing, please?

Conan-Kudo reviewed (approved):
At a glance, things look good to me. I don't have a sample on hand to validate this new logic, but at least it doesn't look broken.
